librz/bin/bobj.c

```diff
@@ -17,20 +17,143 @@ RZ_API void rz_bin_mem_free(void *data) {
 	free(mem);
 }
 
-static int reloc_cmp(const void *a, const RBNode *b, void *user) {
-	const RzBinReloc *ar = (const RzBinReloc *)a;
-	const RzBinReloc *br = container_of(b, const RzBinReloc, vrb);
+static int reloc_cmp(const void *a, const void *b) {
+	const RzBinReloc *ar = a;
+	const RzBinReloc *br = b;
 	if (ar->vaddr > br->vaddr) {
 		return 1;
 	}
 	if (ar->vaddr < br->vaddr) {
 		return -1;
 	}
+	if (ar->paddr > br->paddr) {
+		return 1;
+	}
+	if (ar->paddr < br->paddr) {
+		return -1;
+	}
+	if (ar->type > br->type) {
+		return 1;
+	}
+	if (ar->type < br->type) {
+		return -1;
+	}
+	if (ar->target_vaddr > br->target_vaddr) {
+		return 1;
+	}
+	if (ar->target_vaddr < br->target_vaddr) {
+		return -1;
+	}
 	return 0;
 }
 
-static void reloc_free(RBNode *rbn, void *user) {
-	free(container_of(rbn, RzBinReloc, vrb));
+static int reloc_target_cmp(const void *a, const void *b) {
+	const RzBinReloc *ar = a;
+	const RzBinReloc *br = b;
+	if (ar->target_vaddr > br->target_vaddr) {
+		return 1;
+	}
+	if (ar->target_vaddr < br->target_vaddr) {
+		return -1;
+	}
+	if (ar->vaddr > br->vaddr) {
+		return 1;
+	}
+	if (ar->vaddr < br->vaddr) {
+		return -1;
+	}
+	if (ar->paddr > br->paddr) {
+		return 1;
+	}
+	if (ar->paddr < br->paddr) {
+		return -1;
+	}
+	if (ar->type > br->type) {
+		return 1;
+	}
+	if (ar->type < br->type) {
+		return -1;
+	}
+	return 0;
+}
+
+static RzBinRelocStorage *reloc_storage_new(RZ_OWN RzList *relocs) {
+	RzBinRelocStorage *ret = RZ_NEW0(RzBinRelocStorage);
+	if (!ret) {
+		return NULL;
+	}
+	RzPVector sorter;
+	rz_pvector_init(&sorter, NULL);
+	rz_pvector_reserve(&sorter, rz_list_length(relocs));
+	RzPVector target_sorter;
+	rz_pvector_init(&target_sorter, NULL);
+	rz_pvector_reserve(&target_sorter, rz_list_length(relocs));
+	RzListIter *it;
+	RzBinReloc *reloc;
+	rz_list_foreach (relocs, it, reloc) {
+		rz_pvector_push(&sorter, reloc);
+		if (rz_bin_reloc_has_target(reloc)) {
+			rz_pvector_push(&target_sorter, reloc);
+		}
+	}
+	rz_pvector_sort(&sorter, reloc_cmp);
+	ret->relocs_count = rz_pvector_len(&sorter);
+	ret->relocs = (RzBinReloc **)rz_pvector_flush(&sorter);
+	rz_pvector_fini(&sorter);
+	rz_pvector_sort(&target_sorter, reloc_target_cmp);
+	rz_pvector_shrink(&target_sorter);
+	ret->target_relocs_count = rz_pvector_len(&target_sorter);
+	ret->target_relocs = (RzBinReloc **)rz_pvector_flush(&target_sorter);
```

Comment:
But why?

Reply:
Why what exactly?

Reply:
Why not just use RzPVector?

Reply:
It's not necessary anymore here since the `RzBinRelocStorage` is meant to be immutable. `RzPVector` is primarily useful when you dynamically edit the array.